certora/specs/HealthStatusInvariant.spec

```diff
@@ -0,0 +1,207 @@
+    if (token == ERC20a) {
+        ERC20a.transferFrom(e, from, to, value);
+    } else if (token == EToken) {
+        EToken.transferFrom(e, from, to, value);
+    }
+}
+
+/*
+* The prover struggles to reason about the low-level call operations involved
+* in the real EVCClient.enforceControlCollateral function, so we need
+* to emulate the real behavior here. Here's how it works in the real code:
+*  - EVCClient calls evc.controlCollateral passing a call to `transfer(receiver, amount)` along with the collateral and from addresses
+*  - In controlCollateral the from address is used to set the onBehalfOfAccount
+*  and some authentication is done
+* - After this, callWithContextInternal invokes the transfer function 
+* on the collateral address
+* - Collaterals in the EVK must all be Token.sol and token's transfer
+* implementation calls initOperation which enqueues an account status
+* check on the EVC for the onBehalfOfAccount it also gets from EVC.
+* Because onBehalfOfAccount was set to the from address in callWithContextInternal this status check is for the from account
+* To emulate this, we:
+* - explicitly call EToken.transferFrom using the expected addresses
+* - enqueue a status check on the evc for the "from" address
+*/
+function CVLEnforceCollateralTransfer(env e, address collateral, uint256 amount, address from, address receiver) {
+    // (Ideally: Cast collateral address into EToken to allow for
+    // multiple addresses with the EToken contract)
+    evc.requireAccountStatusCheck(e, from);
+    EToken.transferFromInternalHarnessed(e, from, receiver, amount);
+}
+
+// Assuming the prices stay the same, a healthy account can never become 
+// unhealthy. Here, our assumption that the prices do not change is implicit
+// in the fact that the summary for GetQuote is an uninterpreted function --
+// the prover will model it as a function so it will always return the same
+// value when given the same arguments.
+rule accountsStayHealthy (method f) filtered { f -> 
+    // Literal selectors are used to avoid compilation errors when
+    // only some of the modules are in the verification scene
+    // sig:GovernanceModule.clearLTV(address).selector
+    f.selector != 0x8255d029 && 
+    // sig:GovernanceModule.setLTV(address,uint16,uint16,uint32).selector
+    f.selector != 0x4bca3d5b &&
+    // sig:InitializeModule.initialize(address).selector
+    f.selector != 0xc4d66de8 
+}{
+    env e;
+    calldataarg args;
+    address account;
+    address[] collaterals = evc.getCollaterals(e, account);
+    require collaterals.length == 2; // loop bound
+    require oracleAddress != 0;
+
+    // Vault cannot be a user of itself
+    require account != currentContract;
+    // Vault should not be used as a collateral
+    require collaterals[0] != currentContract;
+    require collaterals[1] != currentContract;
+    // Collaterals must be ETokens
+    require collaterals[0] == EToken;
+    require collaterals[1] == EToken;
```

Comment:
Fixed for all except `Liquidation.liquidate`, which is ever so slightly less general. For all other methods in the entire EVK I allow collateral[0] and collateral[1] to be either ETokenA or ETokenB (if they exist). For `liquidation.liquidate` I fix collateral[0] to be ETokenA and collateral[1] to be ETokenB always (it hits a timeout at 10 hours otherwise). The material difference is that for the way it's done in liquidation.liquidate now, we assume the collaterals are different addresses. However in previous runs of liquidation.liquidate they were forced to be the same address so essentially with the runs combined we covered all cases anyway.